src/Util/HTTPSig.php

```diff
@@ -0,0 +1,400 @@
+					$field = str_replace('_', '-', strtolower(substr($k, 5)));
+					$headers[$field] = $v;
+				}
+			}
+		}
+
+		$sig_block = null;
+
+		if (array_key_exists('signature', $headers)) {
+			$sig_block = self::parseSigheader($headers['signature']);
+		} elseif (array_key_exists('authorization', $headers)) {
+			$sig_block = self::parseSigheader($headers['authorization']);
+		}
+
+		if (!$sig_block) {
+			logger('no signature provided.');
+			return $result;
+		}
+
+		// Warning: This log statement includes binary data
+		// logger('sig_block: ' . print_r($sig_block,true), LOGGER_DATA);
+
+		$result['header_signed'] = true;
+
+		$signed_headers = $sig_block['headers'];
+		if (!$signed_headers) {
+			$signed_headers = ['date'];
+		}
+
+		$signed_data = '';
+		foreach ($signed_headers as $h) {
+			if (array_key_exists($h, $headers)) {
+				$signed_data .= $h . ': ' . $headers[$h] . "\n";
+			}
+			if (strpos($h, '.')) {
+				$spoofable = true;
+			}
+		}
+
+		$signed_data = rtrim($signed_data, "\n");
+
+		$algorithm = null;
+		if ($sig_block['algorithm'] === 'rsa-sha256') {
+			$algorithm = 'sha256';
+		}
+		if ($sig_block['algorithm'] === 'rsa-sha512') {
+			$algorithm = 'sha512';
+		}
+
+		if ($key && function_exists($key)) {
+			$result['signer'] = $sig_block['keyId'];
+			$key = $key($sig_block['keyId']);
+		}
+
+		// We don't use Activity Pub at the moment.
+//		if (!$key) {
+//			$result['signer'] = $sig_block['keyId'];
+//			$key = self::getActivitypubKey($sig_block['keyId']);
+//		}
+
+		if (!$key) {
+			return $result;
+		}
+
+		$x = Crypto::rsaVerify($signed_data, $sig_block['signature'], $key, $algorithm);
+
+		logger('verified: ' . $x, LOGGER_DEBUG);
+
+		if (!$x) {
+			return $result;
+		}
+
+		if (!$spoofable) {
+			$result['header_valid'] = true;
+		}
+
+		if (in_array('digest', $signed_headers)) {
+			$result['content_signed'] = true;
+			$digest = explode('=', $headers['digest']);
+
+			if ($digest[0] === 'SHA-256') {
+				$hashalg = 'sha256';
+			}
+			if ($digest[0] === 'SHA-512') {
+				$hashalg = 'sha512';
+			}
+
+			// The explode operation will have stripped the '=' padding, so compare against unpadded base64.
+			if (rtrim(base64_encode(hash($hashalg, $body, true)), '=') === $digest[1]) {
+				$result['content_valid'] = true;
+			}
+		}
+
+		logger('Content_Valid: ' . $result['content_valid']);
+
+		return $result;
+	}
+
+	/**
+	 * Fetch the public key for Activity Pub contact.
+	 * 
+	 * @param string|int The identifier (contact addr or contact ID).
+	 * @return string|boolean The public key or false on failure.
+	 */
+	private static function getActivitypubKey($id)
+	{
+		if (strpos($id, 'acct:') === 0) {
+			$x = dba::selectFirst('contact', ['pubkey'], ['uid' => 0, 'addr' => str_replace('acct:', '', $id)]);
+		} else {
+			$x = dba::selectFirst('contact', ['pubkey'], ['id' => $id, 'network' => 'activitypub']);
+		}
+
+		if (DBM::is_result($x)) {
+			return $x['pubkey'];
+		}
+
+		if(function_exists('as_fetch')) {
```

Comment:
We don't have a `as_fetch` function in Friendica. You can remove this and everything below.

Reply:
this is about Hubzillas ActivityPub implementation. @annando wanted to keep it so he can work on the Friendica ActivityPub implementation

Reply:
this is about Hubzillas ActivityPub implementation. @annando wanted to keep it so he can work on the Friendica ActivityPub implementation

Reply:
We could leave these parts in some comment or so. I would like to keep it to have a look at it when supporting AP.

Reply:
Since it's ported from Hubzilla anyway, you can always come back and look at it over there, right?